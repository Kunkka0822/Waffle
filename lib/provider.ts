import Ganache, {GanacheOpts} from 'ganache-core';
import {providers, Wallet} from 'ethers';
import defaultAccounts from './config/defaultAccounts';

const defaultGanacheOptions = {accounts: defaultAccounts};

export function createMockProvider(ganacheOptionsOrPathToConfig: string | GanacheOpts = {}) {
  const ganacheOptions = getGanacheOptions(ganacheOptionsOrPathToConfig);
  const options = {...defaultGanacheOptions, ...ganacheOptions};
  return new providers.Web3Provider(Ganache.provider(options));
}

export function getGanacheOptions(ganacheOptionsOrPathToConfig: string | GanacheOpts) {
  if (typeof ganacheOptionsOrPathToConfig === 'object') {
    return ganacheOptionsOrPathToConfig;
  }
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const {ganacheOptions} = require(ganacheOptionsOrPathToConfig);
  return ganacheOptions;
}

export function getWallets(provider: providers.Provider) {
  return defaultAccounts.map((account) => new Wallet(account.secretKey, provider));
}
