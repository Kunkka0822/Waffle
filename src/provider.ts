import Ganache from 'ganache-core';
import {providers, Wallet} from 'ethers';
import defaultAccounts from './config/defaultAccounts';

const defaultGanacheOptions = {accounts: defaultAccounts};

export function createMockProvider(ganacheOptionsOrPathToConfig: string | Ganache.IProviderOptions = {}) {
  const ganacheOptions = getGanacheOptions(ganacheOptionsOrPathToConfig);
  const options = {...defaultGanacheOptions, ...ganacheOptions};
  return new providers.Web3Provider(Ganache.provider(options) as any);
}

export function getGanacheOptions(ganacheOptionsOrPathToConfig: string | Ganache.IProviderOptions) {
  if (typeof ganacheOptionsOrPathToConfig === 'object') {
    return ganacheOptionsOrPathToConfig;
  }
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const {ganacheOptions} = require(ganacheOptionsOrPathToConfig);
  return ganacheOptions;
}

export function getWallets(provider: providers.Provider) {
  return defaultAccounts.map((account) => new Wallet(account.secretKey, provider));
}
