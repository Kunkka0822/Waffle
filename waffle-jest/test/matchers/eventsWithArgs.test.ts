import {BigNumber, Contract, ContractFactory} from 'ethers';
import {MockProvider} from '@ethereum-waffle/provider';
import {EVENTS_ABI, EVENTS_BYTECODE} from '../contracts/Events';

// Note: We make use of Jest Snapshots here because the error output is
// a rich diff with ANSI colors generated by jest-diff and is therefore
// quite large.

describe('INTEGRATION: Events Emitted With Args', () => {
  const [wallet] = new MockProvider().getWallets();
  let factory: ContractFactory;
  let events: Contract;

  beforeEach(async () => {
    factory = new ContractFactory(EVENTS_ABI, EVENTS_BYTECODE, wallet);
    events = await factory.deploy();
  });

  describe('simple', () => {
    it('proper args', async () => {
      const tx = await events.emitOne();
      await expect(tx).toHaveEmittedWith(events, 'One', [
        BigNumber.from(1),
        'One',
        '0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162123'
      ]);
    });

    it('not enough args', async () => {
      const tx = await events.emitOne();
      await expect(
        expect(tx).toHaveEmittedWith(events, 'One', [BigNumber.from(1)])
      ).rejects.toThrowErrorMatchingSnapshot();
      // received additional "One" and address '0x...2123'
    });

    it('too many args', async () => {
      const tx = await events.emitOne();
      await expect(
        expect(tx).toHaveEmittedWith(events, 'One', [1, 2, 3, 4])
      ).rejects.toThrowErrorMatchingSnapshot();
      // received BigNumber, string, and address
    });

    it('one different arg (integer)', async () => {
      const tx = await events.emitOne();
      await expect(
        expect(tx).toHaveEmittedWith(events, 'One', [
          2,
          'One',
          '0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162123'
        ])
      ).rejects.toThrowErrorMatchingSnapshot();
      // expected 2, received BigNumber(1)
    });

    it('one different arg (string number)', async () => {
      const tx = await events.emitOne();
      await expect(
        expect(tx).toHaveEmittedWith(events, 'One', [
          BigNumber.from(1),
          'Two',
          '0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162123'
        ])
      ).rejects.toThrowErrorMatchingSnapshot();
      // expected "Two", received "One"
    });

    it('one different arg (string address)', async () => {
      const tx = await events.emitOne();
      await expect(
        expect(tx).toHaveEmittedWith(events, 'One', [
          BigNumber.from(1),
          'One',
          '0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162124'
        ])
      ).rejects.toThrowErrorMatchingSnapshot();
      // expected '0x...2124', received '0x...2123'
    });
  });

  describe('arrays', () => {
    it('BigNumbers and bytes32', async () => {
      const tx = await events.emitArrays();
      await expect(tx).toHaveEmittedWith(events, 'Arrays', [
        [1, 2, 3].map((x) => BigNumber.from(x)),
        [
          '0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162123',
          '0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162124'
        ]
      ]);
    });

    it('bytes32 ends in different value', async () => {
      const tx = await events.emitArrays();
      await expect(
        expect(tx).toHaveEmittedWith(events, 'Arrays', [
          [1, 2, 3].map((x) => BigNumber.from(x)),
          [
            '0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162121',
            '0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162124'
          ]
        ])
      ).rejects.toThrowErrorMatchingSnapshot();
      // expected '0x...2121', received '0x...2123'
    });

    it('BigNumber received is 1, not 0', async () => {
      const tx = await events.emitArrays();
      await expect(
        expect(tx).toHaveEmittedWith(events, 'Arrays', [
          [0, 2, 3].map((x) => BigNumber.from(x)),
          [
            '0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162123',
            '0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162124'
          ]
        ])
      ).rejects.toThrowErrorMatchingSnapshot();
      // expected "0x00", received "0x01"
    });
  });

  describe('multiples same event', () => {
    it('with different args', async () => {
      const tx = await events.emitOneMultipleTimes();
      await expect(tx).toHaveEmittedWith(events, 'One', [
        BigNumber.from(1),
        'One',
        '0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162123'
      ]);

      await expect(tx).toHaveEmittedWith(events, 'One', [
        BigNumber.from(1),
        'DifferentKindOfOne',
        '0x0000000000000000000000000000000000000000000000000000000000000001'
      ]);
    });

    it('with different args (not found)', async () => {
      const tx = await events.emitOneMultipleTimes();
      await expect(
        expect(tx).toHaveEmittedWith(events, 'One', [1, 2, 3, 4])
      ).rejects.toThrowErrorMatchingSnapshot();
      // received BigNumber(1), "DifferentKindOfOne", "0x...0001"
    });
  });
});
