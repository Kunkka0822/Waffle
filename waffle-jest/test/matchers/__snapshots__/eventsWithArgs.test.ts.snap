// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`INTEGRATION: Events Emitted With Args arrays BigNumber received is 1, not 0 1`] = `
"[32m- Expected[39m
[31m+ Received[39m

[2m  Array [[22m
[2m    Array [[22m
[2m      Object {[22m
[32m-       \\"hex\\": \\"0x00\\",[39m
[31m+       \\"hex\\": \\"0x01\\",[39m
[2m        \\"type\\": \\"BigNumber\\",[22m
[2m      },[22m
[2m      Object {[22m
[2m        \\"hex\\": \\"0x02\\",[22m
[2m        \\"type\\": \\"BigNumber\\",[22m
[2m      },[22m
[2m      Object {[22m
[2m        \\"hex\\": \\"0x03\\",[22m
[2m        \\"type\\": \\"BigNumber\\",[22m
[2m      },[22m
[2m    ],[22m
[2m    Array [[22m
[2m      \\"0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162123\\",[22m
[2m      \\"0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162124\\",[22m
[2m    ],[22m
[2m  ][22m"
`;

exports[`INTEGRATION: Events Emitted With Args arrays bytes32 ends in different value 1`] = `
"[32m- Expected[39m
[31m+ Received[39m

[2m  Array [[22m
[2m    Array [[22m
[2m      Object {[22m
[2m        \\"hex\\": \\"0x01\\",[22m
[2m        \\"type\\": \\"BigNumber\\",[22m
[2m      },[22m
[2m      Object {[22m
[2m        \\"hex\\": \\"0x02\\",[22m
[2m        \\"type\\": \\"BigNumber\\",[22m
[2m      },[22m
[2m      Object {[22m
[2m        \\"hex\\": \\"0x03\\",[22m
[2m        \\"type\\": \\"BigNumber\\",[22m
[2m      },[22m
[2m    ],[22m
[2m    Array [[22m
[32m-     \\"0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162121\\",[39m
[31m+     \\"0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162123\\",[39m
[2m      \\"0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162124\\",[22m
[2m    ],[22m
[2m  ][22m"
`;

exports[`INTEGRATION: Events Emitted With Args multiples same event with different args (not found) 1`] = `
"[32m- Expected[39m
[31m+ Received[39m

[2m  Array [[22m
[32m-   1,[39m
[32m-   2,[39m
[32m-   3,[39m
[32m-   4,[39m
[31m+   Object {[39m
[31m+     \\"hex\\": \\"0x01\\",[39m
[31m+     \\"type\\": \\"BigNumber\\",[39m
[31m+   },[39m
[31m+   \\"DifferentKindOfOne\\",[39m
[31m+   \\"0x0000000000000000000000000000000000000000000000000000000000000001\\",[39m
[2m  ][22m"
`;

exports[`INTEGRATION: Events Emitted With Args simple not enough args 1`] = `
"[32m- Expected[39m
[31m+ Received[39m

[2m  Array [[22m
[2m    Object {[22m
[2m      \\"hex\\": \\"0x01\\",[22m
[2m      \\"type\\": \\"BigNumber\\",[22m
[2m    },[22m
[31m+   \\"One\\",[39m
[31m+   \\"0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162123\\",[39m
[2m  ][22m"
`;

exports[`INTEGRATION: Events Emitted With Args simple one different arg (integer) 1`] = `
"[32m- Expected[39m
[31m+ Received[39m

[2m  Array [[22m
[32m-   2,[39m
[31m+   Object {[39m
[31m+     \\"hex\\": \\"0x01\\",[39m
[31m+     \\"type\\": \\"BigNumber\\",[39m
[31m+   },[39m
[2m    \\"One\\",[22m
[2m    \\"0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162123\\",[22m
[2m  ][22m"
`;

exports[`INTEGRATION: Events Emitted With Args simple one different arg (string address) 1`] = `
"[32m- Expected[39m
[31m+ Received[39m

[2m  Array [[22m
[2m    Object {[22m
[2m      \\"hex\\": \\"0x01\\",[22m
[2m      \\"type\\": \\"BigNumber\\",[22m
[2m    },[22m
[2m    \\"One\\",[22m
[32m-   \\"0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162124\\",[39m
[31m+   \\"0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162123\\",[39m
[2m  ][22m"
`;

exports[`INTEGRATION: Events Emitted With Args simple one different arg (string number) 1`] = `
"[32m- Expected[39m
[31m+ Received[39m

[2m  Array [[22m
[2m    Object {[22m
[2m      \\"hex\\": \\"0x01\\",[22m
[2m      \\"type\\": \\"BigNumber\\",[22m
[2m    },[22m
[32m-   \\"Two\\",[39m
[31m+   \\"One\\",[39m
[2m    \\"0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162123\\",[22m
[2m  ][22m"
`;

exports[`INTEGRATION: Events Emitted With Args simple too many args 1`] = `
"[32m- Expected[39m
[31m+ Received[39m

[2m  Array [[22m
[32m-   1,[39m
[32m-   2,[39m
[32m-   3,[39m
[32m-   4,[39m
[31m+   Object {[39m
[31m+     \\"hex\\": \\"0x01\\",[39m
[31m+     \\"type\\": \\"BigNumber\\",[39m
[31m+   },[39m
[31m+   \\"One\\",[39m
[31m+   \\"0x00cfbbaf7ddb3a1476767101c12a0162e241fbad2a0162e2410cfbbaf7162123\\",[39m
[2m  ][22m"
`;
